src/Translations/Constraint/ConstraintMessageExtractor.php

```diff
@@ -0,0 +1,107 @@
+<?php declare(strict_types=1);
+
+namespace Becklyn\PhpCs\Translations\Constraint;
+
+use Symfony\Component\Validator\Constraint;
+
+/**
+ * Extracts messages from Constraint annotations in Symfony.
+ */
+class ConstraintMessageExtractor
+{
+    /**
+     * Adds all messages from all constraint annotations
+     *
+     * @param object[] $annotations
+     * @return string[]
+     */
+    public function extractMessages (array $annotations) : array
+    {
+        $result = [];
+        foreach ($annotations as $constraint)
+        {
+            if (!$constraint instanceof Constraint)
+            {
+                continue;
+            }
+
+            foreach ($this->extractMessageKeysFromConstraint($constraint) as $message)
+            {
+                $result[$message] = true;
+            }
+        }
+
+        return \array_keys($result);
+    }
+
+
+    /**
+     * @param Constraint $constraint
+     *
+     * @return array
+     */
+    private function extractMessageKeysFromConstraint (Constraint $constraint) : array
+    {
+        $messages = [];
+        $reflectionClass = new \ReflectionClass($constraint);
+
+        // fetch all properties that either are "message" or end in "*Message"
+        foreach ($reflectionClass->getProperties(\ReflectionProperty::IS_PUBLIC) as $property)
+        {
+            if ("message" === $property->getName() || \preg_match('~^.+Message$~', $property->getName()))
+            {
+                $messages = $this->extract($messages, $constraint, $property->getName());
+            }
+        }
+
+        return $messages;
+    }
+
+
+    /**
+     * Extracts the message in the property from the constraint and adds it to the list,
+     * if it is applicable and not empty.
+     *
+     * @param array      $list
+     * @param Constraint $constraint
+     * @param string     $property
+     *
+     * @return array
+     */
+    private function extract (array $list, Constraint $constraint, string $property) : array
+    {
+        if (\property_exists($constraint, $property) && $this->shouldAdd($constraint, $property))
+        {
+            $list[] = $constraint->{$property};
+        }
+
+        return $list;
+    }
+
+
+    /**
+     * Checks whether the property was actually changed in the constraint
```

Comment:
Isn't this potentially a hidden bug? When I'm creating a custom constraint and forget to add the translation for it, it may never be reported. Maybe we should just exclude constraints from the app and only ignore external, if that's easily possible?

Reply:
No, it's not. Otherwise you would get reports for all symfony base constraints, that already have translations.
Basically every external constraint is supposed to bring its own translations for the translation keys it is using.

Fetching the translations for default values of custom constraints is not implemented and would be a different extractor.